p2panda-rs/src/document/document_view_id.rs

```diff
@@ -37,147 +39,113 @@ use super::error::DocumentViewIdError;
 ///                         \
 ///                          \__ [UPDATE] (Hash: "eff..")
 /// ```
-#[derive(Clone, Debug, Eq)]
+#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq, StdHash, Serialize)]
 pub struct DocumentViewId(Vec<OperationId>);
 
 impl DocumentViewId {
     /// Create a new document view id.
-    pub fn new(graph_tips: &[OperationId]) -> Result<Self, DocumentViewIdError> {
-        let document_view_id = Self(graph_tips.to_vec());
-        document_view_id.validate()?;
-        Ok(document_view_id)
-    }
-
-    /// Get the graph tip ids of this view id.
-    pub fn graph_tips(&self) -> &[OperationId] {
-        self.0.as_slice()
-    }
+    ///
+    /// The given operation ids will automatically be sorted and de-duplicated.
+    pub fn new(graph_tips: &[OperationId]) -> Self {
+        let mut graph_tips = graph_tips.to_owned();
 
-    /// Get sorted graph tips for this view id.
-    pub fn sorted(&self) -> Vec<OperationId> {
-        let mut graph_tips = self.0.clone();
+        // Make sure all operations are sorted
         graph_tips.sort();
```

Comment:
I wanted to highlight this change in the behaviour of the `DocumentViewId` constructor. I seem to remember that @sandreae made sure that we did not sort operation ids in document view ids because their order represented information about the graph structure. Do I misremember this? Is that actually not the case?

Reply:
You're looking at a DocumentViewId where indeed the operation ids need to be sorted and de-duplicated. The semantic meaning of arrays you are referring to can be found in relation lists and pinned relation lists :+1:

Reply:
No, I mean `DocumentViewId`!

Reply:
> No, I mean `DocumentViewId`!

Hmn, I do not remember any meaning in their ordering. Don't think there is anything but surely @sandreae knows more.

Reply:
Hmm, I can't remember there being meaning to their ordering either 😅

Reply:
TELL US WHAT YOU SEE SAM!

Reply:
It's hard to piece this together from just pr comments buuut

https://github.com/p2panda/p2panda/pull/277#issuecomment-1073104989

Here I was advocating for sorting document view id hashes in this constructor and we agreed not to do it?

Reply:
> Here I was advocating for sorting document view id hashes in this constructor and we agreed not to do it?

Seems to be a different topic though, this is related to the hashing method and not what we consider a valid document view id (if the sorting has implications for being rejected or not).

Still its an interesting old discussion which also made it back into this PR .. I'm sorting now everything in the constructor when it comes through `new` as doing it yourself before is too complicated. Nice thing is also, that we don't need to return a `Result` then. There is this `from_untrusted` method which is not public which is used by all places where we come from the outside. This one doesn't sort but instead validates the input.

Reply:
That jogs my memory a little. I think my position was coming from implementing `Eq` and that two `DocumentViewId`s which contain the same `OperationId`s should be equal regardless of the order.

> Thanks for the explanations @cafca. For point 1) looking at this from the functionality of a DocumentViewId I would expect two IDs to be equal even if the hashes are listed in different orders. The order has no meaning wrt which document view you are refering too. So the custom eq approach does make sense to me.